crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2041,11 +2042,24 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
 
             (Type::StringLiteral(lhs), Type::StringLiteral(rhs), ast::Operator::Add) => {
-                Type::StringLiteral(StringLiteralType::new(self.db, {
-                    let lhs_value = lhs.value(self.db).to_string();
-                    let rhs_value = rhs.value(self.db).as_ref();
-                    (lhs_value + rhs_value).into()
-                }))
+                let lhs_value = lhs.value(self.db).to_string();
+                let rhs_value = rhs.value(self.db).as_ref();
+                if lhs_value.len() + rhs_value.len() <= Self::MAX_STRING_LITERAL_SIZE {
+                    Type::StringLiteral(StringLiteralType::new(self.db, {
+                        (lhs_value + rhs_value).into()
+                    }))
+                } else {
+                    Type::LiteralString
+                }
+            }
+
+            (
+                Type::StringLiteral(_) | Type::LiteralString,
+                Type::LiteralString,
+                ast::Operator::Add,
+            )
+            | (Type::LiteralString, Type::StringLiteral(_), ast::Operator::Add) => {
```

Comment:
Given that match arms are checked in order, and we already handled `(StringLiteral, StringLiteral, Add)` above, can we simplify this pattern to `((StringLiteral | LiteralString), (StringLiteral | LiteralString), Add)`?

Reply:
Good idea! I guess if someone changed around the order one of the unit tests would fail, so this seems pretty safe.